### PR TITLE
fix(session-replay-browser): add in-flight guard to recordEvents() to prevent concurrent rrweb races

### PR DESCRIPTION
**Pull request by lewgordon-amplitude**
## Why

Session replay files were accumulating large numbers of duplicate `FullSnapshot` events (type=2), inflating file sizes significantly. A real session file showed **22 full snapshots totalling 83% of a 22 MB file** — where a correct recording should produce 1–2.

The root cause: `recordEvents()` is async with three `await` suspension points between `stopRecordingEvents()` (which nulls `recordCancelCallback`) and the final `recordFunction()` call:

```ts
stopRecordingEvents()              // nulls recordCancelCallback synchronously
await getRecordFunction()          // suspension point 1
await initializeNetworkObservers() // suspension point 2
recordCancelCallback = recordFunction({
  plugins: await getRecordingPlugins() // suspension point 3
})
```

During any of these awaits, a concurrent caller (primarily `focusListener` firing while init's `recordEvents()` is mid-await) sees `recordCancelCallback === null`, passes the guard, and both callers invoke rrweb's `record()` independently — each producing a full snapshot.

The race was introduced in May 2025 when `recordEvents()` was first made async for dynamic imports, and widened in June 2025 when rrweb itself was dynamically imported. It was made significantly more visible in April 2026 when SR-3115 added immediate flushing of FullSnapshot events — each duplicate now triggers an immediate network request rather than batching.

## What

Adds a `recordEventsInFlight` boolean guard: concurrent calls to `recordEvents()` return immediately rather than racing through the three await suspension points and spawning a second rrweb recorder instance.

Fixes SR-3531. Follow-up PR #1689 addresses the `focusListener` restart behavior (Root Cause 2).

## Test plan
- [ ] New unit tests: 4 cases covering concurrent call dropped, guard resets on success, guard resets on failure (finally fires), sequential calls both proceed
- [ ] New Playwright e2e: asserts exactly 1 FullSnapshot emitted when focus fires during async init
- [ ] 209/209 unit tests passing, 100% branch coverage

---

> [!NOTE]
> **Medium Risk**
> Changes core recording startup behavior by serializing concurrent `recordEvents()` calls, which could impact edge-case timing around focus/blur or init; mitigated by extensive new unit tests and a new Playwright e2e guard test.
> 
> **Overview**
> Prevents concurrent `recordEvents()` executions from starting multiple rrweb recorders (and emitting duplicate `FullSnapshot` events) by adding an in-flight guard and replaying at most one pending call after the current run completes.
> 
> Refactors the original recording logic into a private `_recordEvents()` helper and updates tests to account for async init/background recording (timer draining and awaiting `setSessionId().promise`). Adds new unit coverage for concurrency/guard reset behavior and a new Playwright e2e spec that asserts `FullSnapshot` counts on init and after a focus-triggered restart (documenting a known limitation).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 55e20a848f4574a63f76614ca0195cb5a965cb5c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by lewgordon-amplitude** [comment]
bugbot run

**Comment by lewgordon-amplitude** [comment]
🚀 **Prerelease published** from RC branch \`sr-3531-rc-2\`

\`\`\`
npm install @amplitude/plugin-session-replay-browser@sr-3531-rc-2
\`\`\`

[Workflow run](https://github.com/amplitude/Amplitude-TypeScript/actions/runs/24670251333)

**Comment by lewgordon-amplitude** [comment]
bugbot run

**Review by stevenchien-amplitude — APPROVED** [review]
22 -> 1-2 is insane. nice.

File: packages/session-replay-browser/src/session-replay.ts [hunk 1]
```diff
@@ -730,6 +733,24 @@ export class SessionReplay implements AmplitudeSessionReplay {
...
+    try {
+      await this._recordEvents(shouldLogMetadata);
+      if (this.recordEventsPending) {
+        this.recordEventsPending = false;
+        await this._recordEvents(shouldLogMetadata);
+      }
+    } finally {
+      this.recordEventsInFlight = false;
+      this.recordEventsPending = false;
+    }
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/session-replay.ts:751]: Fixed in f373087a — changed `if (this.recordEventsPending)` to `while (this.recordEventsPending)` so calls arriving during the replay are also picked up. Added a test covering the three-run case (original + replay + nested pending).
